File: 648-replace-words/replace-words.py
class TrieNode:

  def __init__(self):
    self.children = {}
    self.is_end = False


class Solution:

  def replaceWords(self, dictionary: list[str], sentence: str) -> str:
    root = TrieNode()

    for word in dictionary:
      node = root
      for char in word:
        if char not in node.children:
          node.children[char] = TrieNode()
        node = node.children[char]
      node.is_end = True

    def find_shortest_root(word: str) -> str:
      node = root
      prefix = []
      for char in word:
        if char not in node.children:
          break
        prefix.append(char)
        node = node.children[char]
        if node.is_end:
          return "".join(prefix)
      return word

    words = sentence.split(" ")
    return " ".join(find_shortest_root(w) for w in words)

File: 648-replace-words/replace-words.java
class Solution {
    static class TrieNode {
        TrieNode[] children = new TrieNode[26];
        boolean isEnd = false;
    }

    private final TrieNode root = new TrieNode();

    private void insert(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (node.children[idx] == null) {
                node.children[idx] = new TrieNode();
            }
            node = node.children[idx];
        }
        node.isEnd = true;
    }

    private String findShortestRoot(String word) {
        TrieNode node = root;
        StringBuilder prefix = new StringBuilder();

        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (node.children[idx] == null) {
                break;
            }
            prefix.append(c);
            node = node.children[idx];
            if (node.isEnd) {
                return prefix.toString();
            }
        }
        return word;
    }

    public String replaceWords(List<String> dictionary, String sentence) {
        for (String rootWord : dictionary) {
            insert(rootWord);
        }

        String[] words = sentence.split(" ");
        for (int i = 0; i < words.length; i++) {
            words[i] = findShortestRoot(words[i]);
        }

        return String.join(" ", words);
    }
}